tsdb/engine/tsm1/batch_float.go

```diff
@@ -252,222 +253,262 @@ func FloatArrayEncodeAll(src []float64, b []byte) ([]byte, error) {
 	return b[:length], nil
 }
 
-func FloatArrayDecodeAll(b []byte, dst []float64) ([]float64, error) {
-	if len(b) == 0 {
-		return []float64{}, nil
+// bitMask contains a lookup table where the index is the number of bits
+// and the value is a mask. The table is always read by ANDing the index
+// with 0x3f, such that if the index is 64, position 0 will be read, which
+// is a 0xffffffffffffffff, thus returning all bits.
+//
+// 00 = 0xffffffffffffffff
+// 01 = 0x0000000000000001
+// 02 = 0x0000000000000003
+// 03 = 0x0000000000000007
+// ...
+// 62 = 0x3fffffffffffffff
+// 63 = 0x7fffffffffffffff
+var bitMask [64]uint64
+
+func init() {
+	v := uint64(1)
+	for i := 1; i <= 64; i++ {
+		bitMask[i&0x3f] = v
+		v = v<<1 | 1
 	}
+}
 
-	sz := cap(dst)
-	if sz == 0 {
-		sz = 64
-		dst = make([]float64, sz)
-	} else {
-		dst = dst[:sz]
+func FloatArrayDecodeAll(b []byte, buf []float64) ([]float64, error) {
+	if len(b) < 9 {
+		return []float64{}, nil
 	}
 
 	var (
-		val      uint64 // current value
-		leading  uint64
-		trailing uint64
-		bit      bool
-		br       BatchBitReader
+		val         uint64      // current value
+		trailingN   uint8       // trailing zero count
+		meaningfulN uint8  = 64 // meaningful bit count
 	)
 
-	j := 0
+	// first byte is the compression type; always Gorilla
+	b = b[1:]
 
-	// first byte is the compression type.
-	// we currently just have gorilla compression.
-	br.Reset(b[1:])
-	val = br.ReadBits(64)
+	val = binary.BigEndian.Uint64(b)
 	if val == uvnan {
+		if buf == nil {
+			var tmp [1]float64
+			buf = tmp[:0]
+		}
 		// special case: there were no values to decode
-		return dst[:0], nil
+		return buf[:0], nil
 	}
 
-	dst[j] = math.Float64frombits(val)
-	j++
+	buf = buf[:0]
+	// convert the []float64 to []uint64 to avoid calling math.Float64Frombits,
+	// which results in unnecessary moves between Xn registers before moving
+	// the value into the float64 slice. This change increased performance from
+	// 320 MB/s to 340 MB/s on an Intel(R) Core(TM) i7-6920HQ CPU @ 2.90GHz
+	dst := *(*[]uint64)(unsafe.Pointer(&buf))
+	dst = append(dst, val)
+
+	b = b[8:]
+
+	// The bit reader code uses brCachedVal to store up to the next 8 bytes
+	// of MSB data read from b. brValidBits stores the number of remaining unread
+	// bits starting from the MSB. Before N bits are read from brCachedVal,
+	// they are left-rotated N bits, such that they end up in the left-most position.
+	// Using bits.RotateLeft64 results in a single instruction on many CPU architectures.
+	// This approach permits simple tests, such as for the two control bits:
+	//
+	//    brCachedVal&1 > 0
+	//
+	// The alternative was to leave brCachedValue alone and perform shifts and
+	// masks to read specific bits. The original approach looked like the
+	// following:
+	//
+	//    brCachedVal&(1<<(brValidBits&0x3f)) > 0
+	//
+	var (
```

Comment:
@e-dard / @zeebo note the comment regarding the implementation of the bit reading code